test/integration/replicaset/replicaset_test.go

```diff
@@ -515,3 +538,569 @@ func TestRSSelectorImmutability(t *testing.T) {
+		t.Errorf("len(pods) = %v, want %v", len(pods.Items), 3)
+	}
+
+	// Expect both RS have status.replicas = spec.replicas
+	for i := 0; i < 2; i++ {
+		replicaset, err := c.ExtensionsV1beta1().ReplicaSets(ns.Name).Get(fmt.Sprintf("rs-%d", i+1), metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("failed to get replicaset rs-%d: %v", i+1, err)
+		}
+		if replicaset.Status.Replicas != *replicaset.Spec.Replicas {
+			t.Fatalf(".status.replicas %d is not equal to .spec.replicas %d", replicaset.Status.Replicas, *replicaset.Spec.Replicas)
+		}
+	}
+}
+
+func TestPodOrphaningAndAdoptionWhenLabelsChange(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-replicaset-basics", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+	labelMap := map[string]string{"foo": "bar"}
+	rs := newRS("rs", ns.Name, 1)
+	rs.Spec.Selector.MatchLabels = labelMap
+	rs.Spec.Template.Labels = labelMap
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{}, ns.Name)
+
+	stopCh := make(chan struct{})
+	informers.Start(stopCh)
+	go rm.Run(5, stopCh)
+	defer close(stopCh)
+	waitRSStable(t, c, rs, ns.Name)
+
+	// Orphaning: RS should remove OwnerReference from a pod when the pod's labels change to not match its labels
+	// Although we can access the pod directly, we should still access it via selector to ensure the selector is working
+	podClient := c.Core().Pods(ns.Name)
+	podSelector := labels.Set{"foo": "bar"}.AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := podClient.List(options)
+	if err != nil {
+		t.Fatalf("Failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
+	}
+	if pods == nil {
+		t.Fatalf("Obtained a nil list of pods")
+	}
+	if len(pods.Items) != 1 {
+		t.Fatalf("len(pods) = %v, want %v", len(pods.Items), 1)
+	}
+
+	pod := &pods.Items[0]
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		newPod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		// Firstly, test ControllerRef for the pod is not nil
+		if metav1.GetControllerOf(newPod) != nil {
+			pod = newPod
+			return true, nil
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	newLabelMap := map[string]string{"new-foo": "new-bar"}
+	if err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		pod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		pod.Labels = newLabelMap
+		pod, err = podClient.Update(pod)
+		if err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		newPod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		// Then, test ControllerRef for the pod is nil
+		if metav1.GetControllerOf(newPod) == nil {
+			pod = newPod
```

Comment:
You're not using this value anywhere. You only needed this for the last case, because you make use of `pod` after the function returns.

Reply:
`pod` is still used by next `RetryOnConflict` call. Do you mean a global value initiated before callback but altered during callback will not have its updated value reflected after callback returns?
For example:
A := 1
callback {
  A = 2
}
//A is still 1

Reply:
A global value initiated before callback but altered during callback will NOT have its updated value reflected after callback returns. I found it out via this: https://play.golang.org/p/pml9xby1Ya

I will change all affected codes.

Reply:
The Go playground example you gave is a different situation. That one just shows that variables are passed by value. Go does not have pass by reference.

A more relevant example for this case would be this: https://play.golang.org/p/BTrpWXfOfG

Reply:
My understanding was wrong - Golang's (and other languages') callback functions will change the value after the callback functions return. It make sense why you suggest `pod = newPod` now.
For code readability, i will:
(1) use `newPod` and/or `newRS` in all callback functions
(2) use `t.Fatalf()` to print all error messages